File: dotnet-engine/Yggdrasil.Engine/Strategies.cs
namespace Yggdrasil;

using System;
using System.Text.Json;

internal class CustomStrategies
{
    private JsonSerializerOptions options = new JsonSerializerOptions
    {
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase
    };

    private string[]? knownStrategies = null;

    private Dictionary<string, IStrategy> strategies = new Dictionary<string, IStrategy>();
    private Dictionary<string, MappedFeature>? mappedFeatures = null;

    internal CustomStrategies(string[]? knownStrategies)
    {
        this.knownStrategies = knownStrategies;
    }

    private bool IsCustomStrategy(StrategyDefinition strategy)
    {
        return !knownStrategies?.Contains(strategy.Name) ?? false;
    }

    private List<MappedStrategy> MapCustomStrategies(List<StrategyDefinition>? strategies)
    {
        if (strategies == null)
        {
            return new List<MappedStrategy>();
        }

        return strategies
            .Where(IsCustomStrategy)
            .Where(definition => this.strategies?.ContainsKey(definition.Name) ?? false)
            .Select((definition, index) =>
                new MappedStrategy(
                    index,
                    definition.Name,
                    definition.Parameters ?? new Dictionary<string, string>(),
                    this.strategies[definition.Name]))
            .ToList();
    }

    internal void MapFeatures(string json)
    {
        var features = JsonSerializer.Deserialize<FeatureCollection>(json, options)?.Features;
        mappedFeatures = features?
            .Select(feature => new MappedFeature(feature, MapCustomStrategies(feature.Strategies)))
            .ToDictionary(feature => feature.Name);
    }

    internal void RegisterCustomStrategies(List<IStrategy> strategies)
    {
        this.strategies = strategies.ToDictionary(strategy => strategy.Name);
    }

    internal string GetCustomStrategyPayload(string toggleName, Context context)
    {
        MappedFeature? feature = null;
        mappedFeatures?.TryGetValue(toggleName, out feature);
        if (feature == null)
        {
            return "{}";
        }

        var strategies = feature.Strategies
            .ToDictionary(strategy => strategy.ResultName,
                strategy => strategy.IsEnabled(context));

        return JsonSerializer.Serialize(strategies, options);
    }
}

File: javascript-engine/src/strategies.ts
import type {
  Context,
  DeltaState,
  Feature,
  State,
  Strategy,
  StrategyDefinition
} from './types'

type MappedStrategy = {
  resultName: string
  strategyName: string
  strategy: Strategy
  parameters: Record<string, string>
  isEnabled: (context: Context) => boolean
}

type MappedFeature = {
  name: string
  strategies: MappedStrategy[]
}

export class Strategies {
  private knownStrategies: string[]
  private strategies: Map<string, Strategy> = new Map()
  private mappedFeatures: Map<string, MappedFeature> = new Map()

  constructor(knownStrategies: string[] = []) {
    this.knownStrategies = knownStrategies
  }

  private isCustomStrategy(strategyName: string): boolean {
    return !this.knownStrategies.includes(strategyName)
  }

  private mapCustomStrategies(
    strategies?: StrategyDefinition[]
  ): MappedStrategy[] {
    if (!strategies) {
      return []
    }

    return strategies
      .filter(
        ({ name }) => this.strategies.has(name) && this.isCustomStrategy(name)
      )
      .map((definition, index) => {
        const strategy = this.strategies.get(definition.name)!
        const parameters = definition.parameters ?? {}

        return {
          resultName: `customStrategy${index + 1}`,
          strategyName: definition.name,
          parameters,
          strategy,
          isEnabled: (context: Context) =>
            strategy.isEnabled(parameters, context)
        }
      })
  }

  private mapFeature(feature: Feature): MappedFeature {
    return {
      name: feature.name,
      strategies: this.mapCustomStrategies(feature.strategies)
    }
  }

  mapFeatures(state: State | DeltaState): void {
    if ('features' in state) {
      this.mappedFeatures = new Map(
        state.features.map(feature => [feature.name, this.mapFeature(feature)])
      )
      return
    }

    if ('events' in state) {
      state.events
        .sort((a, b) => a.eventId - b.eventId)
        .forEach(event => {
          if (event.type === 'hydration') {
            event.features.forEach(feature =>
              this.mappedFeatures.set(feature.name, this.mapFeature(feature))
            )
          }

          if (event.type === 'feature-updated') {
            this.mappedFeatures.set(
              event.feature.name,
              this.mapFeature(event.feature)
            )
          }

          if (event.type === 'feature-removed') {
            this.mappedFeatures.delete(event.featureName)
          }
        })
    }
  }

  registerCustomStrategies(strategies: Strategy[]): void {
    this.strategies = new Map(
      strategies.map(strategy => [strategy.name, strategy])
    )
  }

  getCustomStrategyPayload(toggleName: string, context: Context) {
    const feature = this.mappedFeatures.get(toggleName)
    if (!feature) {
      return {}
    }

    const strategies = Object.fromEntries(
      feature.strategies.map(strategy => [
        strategy.resultName,
        strategy.isEnabled(context)
      ])
    )

    return strategies
  }
}
